pageserver/src/repository/rocksdb.rs

```diff
@@ -90,6 +89,38 @@ struct CacheKey {
     pub lsn: Lsn,
 }
 
+//
+// In addition to those per-page entries, the 'last_valid_lsn' and 'last_record_lsn'
+// values are also persisted in the rocskdb repository. They are stored with CacheKeys
+// with ROCKSDB_SPECIAL_FORKNUM, and 'blknum' indicates which value it is. The
+// rest of the key fields are zero. We use a CacheKey as the key for these too,
+// so that whenever we iterate through keys in the repository, we can safely parse
+// the key blob as CacheKey without checking for these special values first.
+//
+impl CacheKey {
```

Comment:
@knizhnik so I used these special keys to store the "last_valid_lsn" and "last_record_lsn" in the rocksdb repository. Meanwhile, you added the `fork` function to also create similar magic entries in the repository. As we just discussed, we really need to clean up the way we store all these non-relation things and the keys we use for them

Reply:
Some my thoughts concerning subj:
1. Using dummy forknumber for storing  non-relational data really seems to be a kind of "dirty hack" solution. May be we should add some independent "kind" field to the key. But using "synthetic" for numbers for pg_xact/pg_multixact allows to reuse existed implementation and avoid code duplication. 
2. It is not clear if storing relational and non-relational data in the same repository (file) is really good idea. Non relational data is larger and can not be traversed in reasonable amount of time. Non-relational data is small. To be able to efficiently extract relfilemap  nodes I have to change order of sorting and place forknum key first. If each kid of data will be stored in separate database (file), there will be non need in such hacks and  small non-relational data wil be extracted faster. In any case, sch details should be hidden in storage implementation.
3. Right now we do not have abstraction of storage manager or in other words - persistent engine implementation. It is encapsulated in Timeline trait. But it seems to be confusing and unnatural to tell about Rocksdb timelines. So as far as we are refactoring API in any case, may be it is better to add trait PersistentEngine or PersistentStorage and provide its implementation for RocksDB&Co? In this case timeline code will just create engine instance (using some factory) and use it to implement timeline functionality?

Reply:
> Some my thoughts concerning subj:
> 
>     1. Using dummy forknumber for storing  non-relational data really seems to be a kind of "dirty hack" solution. May be we should add some independent "kind" field to the key. But using "synthetic" for numbers for pg_xact/pg_multixact allows to reuse existed implementation and avoid code duplication.

Yeah, we should refactor that. Also discussed at https://github.com/zenithdb/zenith/pull/174. I think we can handle that as part of that PR, and keep it as it is here.

>     2. It is not clear if storing relational and non-relational data in the same repository (file) is really good idea. Non relational data is larger and can not be traversed in reasonable amount of time. Non-relational data is small. To be able to efficiently extract relfilemap  nodes I have to change order of sorting and place forknum key first. If each kid of data will be stored in separate database (file), there will be non need in such hacks and  small non-relational data wil be extracted faster. In any case, sch details should be hidden in storage implementation.

Agreed, that's an implementation detail in the Repository/Timeline impementation.

>     3. Right now we do not have abstraction of storage manager or in other words - persistent engine implementation. It is encapsulated in Timeline trait. But it seems to be confusing and unnatural to tell about Rocksdb timelines. So as far as we are refactoring API in any case, may be it is better to add trait PersistentEngine or PersistentStorage and provide its implementation for RocksDB&Co? In this case timeline code will just create engine instance (using some factory) and use it to implement timeline functionality?

Hmm, so the Repository/Timeline interface would be renamed to PersistentEngine, and would only deal with a single timeline? And the code manage multiple timelines would be at a layer above that?

I don't think that's a good idea, the storage engine really should know about timelines, for an efficient implementation. For example, if timeline B is forked off from timeline B, you only want to store the pages that were modified on timeline B. That's straightforward, but it gets harder when consider garbage collecting the old page versions from timeline A, if timeline A is no longer needed. That kind of things seem very hard to abstract. I think dealing with garbage collection and efficiently storing differences between timelines is a core functionality of the Repository/Timeline implementation.